### DISCUSSION TITLE
Inconsistent margin for 2D Barcodes

### DISCUSSION THREAD

**Discussion by mbaros100** [zxing-cpp/zxing-cpp#482] [Q&A]
2D barcodes sometimes generated with the margin i specified, sometimes it is way too big depending on the specified size. 
I do understand this might be done to preserve the proportional ratio of the 2d barcode.

I noticed that for my specific barcode i was experimenting with it perfects the results on specific sizes... like 63, 84, etc. if you specify any size between 63 and 84, it is going to generate a barcode with size 63x63 and use the other space to do extra padding.

I want to know if this is as designed or we can do something about it in order to get a margin that was exactly specified by me?

Why not simply resizing the result to the size we specified in the input instead of doing extra padding? Are there cases this scenario will result in non-working barcode? If so, can i have an example please?

Thanks a lot.

Attached is a QR that i tired to generate with 0 margin and 80 width/height with data=123456789

![qr](https://user-images.githubusercontent.com/8537079/211681938-b5a3be77-88c2-46c1-8438-46df023fc779.png)

**Comment by axxel** [comment]
Yes, this behavior is as designed. The process to generate a valid QR Code symbol is basically:
1. depending on the input length and the error correction level, one can calculate the minimum version (meaning size) of the symbol, say 21 modules as in your case. To be strictly conformant there need be to at least 4 modules of white space around the symbol.
2. the library generates a `BitMatrix` with the minimum size where each 'pixel' corresponds to exactly one module
3. if the user wants something bigger, say 80 pixels, it can scale the symbol up 3x (21x3 = 63) and fill the rest with white. 4x (=84) is obviously too big.

If you want a pure black/white output that is still conforming to the specification, meaning 1 module is N pixels wide across the whole image, then this is the only option. If you would render those 21 modules into 28 pixels, say, you would need to add 7 pixels across the symbol, so 7 rows/columns would be 2 pixels wide, while all the others would be 1 pixel wide. This completely destroys the readability of the symbol. You end up with problems like the one discussed in #479 and [here](https://github.com/zxing-cpp/zxing-cpp/issues/454#issuecomment-1338796691).

If you scale that up 10x, so N would be 10, but you wanted to have exactly 220 pixels instead of 210, then this problem would in fact 'disappear' because then 10 rows/columns would be 11 instead of 10 pixels wide but that is only a 10% error and that would certainly work. Where exactly the threshold is, I don't know and it will depend on the used symbology. @gitlost I wander how zint is dealing with this issue while making sure the result is always a valid symbol?

If you need some exact size, I would suggest the following:
1. render your symbol to the minimal size
2. scale it up to whatever size you need with an external library/algorithm but make sure that a) your new size is at least 10x the original or b) render it to a greyscale image with a bilinear or bicubic scaling method.

**Comment by gitlost** [comment]
Hi axxel, it doesn't! in the sense that zint will let you produce a raster symbol with varying module size using interpolation if that's what you wish, although it's not recommended.

To produce a specific pixel size you have to do some maths as zint uses a `--scale` option to size barcodes. Here `80 / 21 = ~3.8095`, and zint's scaling units are 2 pixels so you need to halve this and specify `1.91`, rounding up:
```
zint -b QRCODE -d 123456789 --secure=1 --scale=1.91
```
This produces an 80x80 PNG with most rows/columns 4 pixels wide and some rows/columns 3 pixels wide:

![out](https://user-images.githubusercontent.com/481982/211693101-4cd79ca1-972c-47f6-9159-a9e40ef98ff6.png)

which actually the truly wonderful ZXing-C++ will scan ok.

Obviously this is a generally a bad idea as you say, and zint recommends half-integer scaling only for raster output. (Margins or "whitespace" in zint terminology are specified separately with the options `--whitespace`, `--vwhitespace` and `--quietzones`.)

**Reply by axxel** [reply]
@gitlost thanks for the detailed feedback. Followup questions:
1. So that means a sale of 1 results in 1 module == 2 pixels? And one can use `--scale=0.5`? Is there a technical reason for this choice?
2. If someone wants to have a certain image size as output they have to call zint twice then: first find out how large the symbol is going to be by default and then calculate the scale value based on that and call it again, right?
3. Has the option of using a bilinear scaling and outputting greyscale png files ever been considered?

**Reply by gitlost** [reply]
Re 1. scale 1 == 2 pixels, no, there's no technical reason, it's historical, kept for back-compatibility.

Re 2. if one's using the library, the `scale` is only used on printing the symbol, so one could call the encode function first and then using that info set the scale accordingly before calling the print function.

Re 3. bilinear scaling, do you mean separate x / y scale factors? If so, yes this has been briefly considered for linear and stacked barcodes, and might be introduced if someone finds it useful, but there's a `height` option that can be used in conjunction with `scale` that covers most use cases, if awkwardly.

Re 3. greyscale png, no, this hasn't been considered up to now. However I'm hoping to introduce greyscale font rendering in the next release, so this may become an option assuming colour isn't used elsewhere. Any advice on this or on whether using greyscale could be advantageous anyway would be most welcome.

**Reply by gitlost** [reply]
Ha just googled "bilinear scaling" and see that you probably mean "bilinear interpolation", producing greyscale intermediates? No this hasn't been considered before but may be the thing to do once greyscale fonts are implemented, if I understand you correctly.

**Reply by axxel** [reply]
Yes, scaling with bilinear interpolation is what I meant. For font rendering this will certainly be a huge advantage for readability. For the decodability of the barcode itself via zxing it would in fact not make a difference as of today. But that might change in the future. For linear codes it would allow to read the width of the bars and spaces with sub-pixel accuracy. That would then theoretically allow to render a 50 module wide code into say 60 pixels without outright destroying the content.

That said, I believe the users should always be strongly discouraged to use module sizes that are not integers.
